File: packages/prism-core/src/kernel/builder/profiles.ts
/**
 * Built-in App Profiles for the self-replicating Studio.
 *
 * These map the four ecosystem apps (Flux, Lattice, Cadence, Grip) plus
 * the universal Studio host and a Relay-only profile into pinned
 * configurations the BuilderManager can consume.
 */

import type { AppProfile, BuiltInProfileId } from "./types.js";

export const STUDIO_PROFILE: AppProfile = {
  id: "studio",
  name: "Prism Studio",
  version: "0.1.0",
  // undefined plugins/lenses = all built-ins = universal host
  defaultLens: "editor",
  allowGlassFlip: true,
  theme: {
    displayName: "Prism Studio",
  },
  starterApp: {
    label: "Studio",
    description: "Universal host — all lenses, minimal chrome.",
    appShell: {
      data: {
        brand: "Prism Studio",
        brandIcon: "\u25A6",
        topBarHeight: 48,
        leftBarWidth: 220,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: true,
      },
      children: [
        { type: "heading", slot: "topBar", name: "Brand", data: { text: "Prism Studio", level: "h3", align: "left" } },
        { type: "site-nav", slot: "leftBar", name: "Site nav", data: { orientation: "vertical", source: "pages" } },
      ],
    },
    defaultPageShell: {
      data: {
        topBarHeight: 0,
        leftBarWidth: 0,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: false,
      },
    },
    routes: [
      { path: "/", label: "Home", pageTemplate: "landing", isHome: true, showInNav: true },
      { path: "/docs", label: "Docs", pageTemplate: "blog", showInNav: true },
    ],
  },
};

export const FLUX_PROFILE: AppProfile = {
  id: "flux",
  name: "Flux",
  version: "0.1.0",
  plugins: ["work", "finance", "crm"],
  lenses: [
    "editor",
    "canvas",
    "record-browser",
    "automation",
    "work",
    "finance",
    "crm",
  ],
  defaultLens: "record-browser",
  allowGlassFlip: true,
  theme: {
    primary: "#6C5CE7",
    displayName: "Flux",
    brandIcon: "flux.svg",
  },
  kbarCommands: ["new-task", "new-invoice", "new-contact", "start-timer"],
  starterApp: {
    label: "Flux",
    description: "Productivity — tasks, contacts, invoices in a record-browser chrome.",
    appShell: {
      data: {
        brand: "Flux",
        brandIcon: "\u26A1",
        topBarHeight: 56,
        leftBarWidth: 280,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: true,
      },
      children: [
        { type: "heading", slot: "topBar", name: "Brand", data: { text: "Flux", level: "h2", align: "left" } },
        { type: "site-nav", slot: "leftBar", name: "Record nav", data: { orientation: "vertical", source: "pages", showIcons: true } },
      ],
    },
    defaultPageShell: {
      data: {
        topBarHeight: 0,
        leftBarWidth: 0,
        rightBarWidth: 320,
        bottomBarHeight: 0,
        stickyTopBar: false,
      },
    },
    routes: [
      { path: "/", label: "Inbox", pageTemplate: "blank", isHome: true, showInNav: true },
      { path: "/tasks", label: "Tasks", pageTemplate: "blank", showInNav: true },
      { path: "/contacts", label: "Contacts", pageTemplate: "blank", showInNav: true },
      { path: "/invoices", label: "Invoices", pageTemplate: "blank", showInNav: true },
    ],
  },
};

export const LATTICE_PROFILE: AppProfile = {
  id: "lattice",
  name: "Lattice",
  version: "0.1.0",
  plugins: ["assets", "platform"],
  lenses: [
    "editor",
    "graph",
    "spatial-canvas",
    "visual-script",
    "luau-facet",
    "assets-mgmt",
  ],
  defaultLens: "graph",
  allowGlassFlip: true,
  theme: {
    primary: "#00B894",
    displayName: "Lattice",
    brandIcon: "lattice.svg",
  },
  kbarCommands: ["new-dialogue", "compile-bank", "open-entity"],
  starterApp: {
    label: "Lattice",
    description: "Game middleware — graph-first editor with minimal chrome.",
    appShell: {
      data: {
        brand: "Lattice",
        brandIcon: "\u2B22",
        topBarHeight: 40,
        leftBarWidth: 0,
        rightBarWidth: 300,
        bottomBarHeight: 120,
        stickyTopBar: true,
      },
      children: [
        { type: "heading", slot: "topBar", name: "Brand", data: { text: "Lattice", level: "h3", align: "left" } },
      ],
    },
    defaultPageShell: {
      data: {
        topBarHeight: 0,
        leftBarWidth: 0,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: false,
      },
    },
    routes: [
      { path: "/", label: "Graph", pageTemplate: "blank", isHome: true, showInNav: true },
      { path: "/assets", label: "Assets", pageTemplate: "blank", showInNav: true },
      { path: "/scripts", label: "Scripts", pageTemplate: "blank", showInNav: true },
    ],
  },
};

export const CADENCE_PROFILE: AppProfile = {
  id: "cadence",
  name: "Cadence",
  version: "0.1.0",
  plugins: ["life", "platform"],
  lenses: [
    "editor",
    "canvas",
    "luau-facet",
    "record-browser",
    "life",
  ],
  defaultLens: "canvas",
  allowGlassFlip: true,
  theme: {
    primary: "#FD79A8",
    displayName: "Cadence",
    brandIcon: "cadence.svg",
  },
  kbarCommands: ["new-lesson", "transcribe-session", "open-course"],
  starterApp: {
    label: "Cadence",
    description: "Music education — lesson canvas with hero landing.",
    appShell: {
      data: {
        brand: "Cadence",
        brandIcon: "\u266B",
        topBarHeight: 72,
        leftBarWidth: 0,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: true,
      },
      children: [
        { type: "heading", slot: "topBar", name: "Brand", data: { text: "Cadence", level: "h1", align: "center" } },
      ],
    },
    defaultPageShell: {
      data: {
        topBarHeight: 0,
        leftBarWidth: 0,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: false,
      },
    },
    routes: [
      { path: "/", label: "Welcome", pageTemplate: "landing", isHome: true, showInNav: true },
      { path: "/lessons", label: "Lessons", pageTemplate: "blank", showInNav: true },
      { path: "/courses", label: "Courses", pageTemplate: "blank", showInNav: true },
      { path: "/about", label: "About", pageTemplate: "blog", showInNav: true },
    ],
  },
};

export const GRIP_PROFILE: AppProfile = {
  id: "grip",
  name: "Grip",
  version: "0.1.0",
  plugins: ["work", "assets", "platform"],
  lenses: [
    "editor",
    "graph",
    "spatial-canvas",
    "canvas",
    "automation",
    "work",
    "assets-mgmt",
  ],
  defaultLens: "spatial-canvas",
  allowGlassFlip: true,
  theme: {
    primary: "#E17055",
    displayName: "Grip",
    brandIcon: "grip.svg",
  },
  kbarCommands: ["new-cue", "open-stage-plot", "arm-transport"],
  starterApp: {
    label: "Grip",
    description: "Live production — cue list driven spatial canvas.",
    appShell: {
      data: {
        brand: "Grip",
        brandIcon: "\u2726",
        topBarHeight: 48,
        leftBarWidth: 320,
        rightBarWidth: 0,
        bottomBarHeight: 160,
        stickyTopBar: true,
      },
      children: [
        { type: "heading", slot: "topBar", name: "Brand", data: { text: "Grip", level: "h2", align: "left" } },
      ],
    },
    defaultPageShell: {
      data: {
        topBarHeight: 0,
        leftBarWidth: 0,
        rightBarWidth: 0,
        bottomBarHeight: 0,
        stickyTopBar: false,
      },
    },
    routes: [
      { path: "/", label: "Stage", pageTemplate: "blank", isHome: true, showInNav: true },
      { path: "/cues", label: "Cues", pageTemplate: "blank", showInNav: true },
      { path: "/assets", label: "Assets", pageTemplate: "blank", showInNav: true },
    ],
  },
};

export const RELAY_PROFILE: AppProfile = {
  id: "relay",
  name: "Prism Relay",
  version: "0.1.0",
  // Relay has no plugins or lenses — it's a headless server profile.
  plugins: [],
  lenses: [],
  relayModules: [
    "blind-mailbox",
    "relay-router",
    "relay-timestamp",
    "capability-tokens",
    "sovereign-portals",
    "webrtc-signaling",
  ],
  theme: { displayName: "Prism Relay" },
  allowGlassFlip: false,
};

export const BUILT_IN_PROFILES: Record<BuiltInProfileId, AppProfile> = {
  studio: STUDIO_PROFILE,
  flux: FLUX_PROFILE,
  lattice: LATTICE_PROFILE,
  cadence: CADENCE_PROFILE,
  grip: GRIP_PROFILE,
  relay: RELAY_PROFILE,
};

export function listBuiltInProfiles(): AppProfile[] {
  return Object.values(BUILT_IN_PROFILES);
}

export function getBuiltInProfile(id: BuiltInProfileId): AppProfile {
  return BUILT_IN_PROFILES[id];
}

/** Serialize a profile to canonical `.prism-app.json` text. */
export function serializeAppProfile(profile: AppProfile): string {
  return `${JSON.stringify(profile, null, 2)}\n`;
}

/** Parse a `.prism-app.json` file. Throws on invalid shape. */
export function parseAppProfile(source: string): AppProfile {
  const parsed: unknown = JSON.parse(source);
  if (!parsed || typeof parsed !== "object") {
    throw new Error("App profile must be a JSON object");
  }
  const obj = parsed as Record<string, unknown>;
  if (typeof obj.id !== "string" || typeof obj.name !== "string" || typeof obj.version !== "string") {
    throw new Error("App profile requires id/name/version strings");
  }
  return parsed as AppProfile;
}

File: packages/prism-core/src/kernel/builder/profiles.rs
//! `kernel::builder::profiles` — built-in `AppProfile`s for the
//! self-replicating Studio, plus canonical JSON encode/decode.
//!
//! Port of `kernel/builder/profiles.ts` at 8426588. Each of the four
//! ecosystem apps (Flux, Lattice, Cadence, Grip) has a pinned profile;
//! a universal Studio host profile and a headless Relay profile round
//! out the set.

use serde_json::{json, Value as JsonValue};
use std::collections::BTreeMap;

use super::types::{
    AppProfile, AppThemeConfig, BuiltInProfileId, PageTemplateKind, StarterAppTemplate,
    StarterRouteTemplate, StarterShellChild, StarterShellSlot,
};

fn shell_data<const N: usize>(pairs: [(&str, JsonValue); N]) -> BTreeMap<String, JsonValue> {
    pairs.into_iter().map(|(k, v)| (k.into(), v)).collect()
}

fn shell_child(type_name: &str, slot: &str, name: &str, data: JsonValue) -> StarterShellChild {
    StarterShellChild {
        type_name: type_name.into(),
        slot: slot.into(),
        name: Some(name.into()),
        data: data.as_object().map(|o| {
            o.iter()
                .map(|(k, v)| (k.clone(), v.clone()))
                .collect::<BTreeMap<_, _>>()
        }),
    }
}

fn route(
    path: &str,
    label: &str,
    template: PageTemplateKind,
    is_home: bool,
) -> StarterRouteTemplate {
    StarterRouteTemplate {
        path: path.into(),
        label: label.into(),
        page_template: template,
        is_home: if is_home { Some(true) } else { None },
        show_in_nav: Some(true),
    }
}

// ── Built-in profiles ───────────────────────────────────────────────────────

pub fn studio_profile() -> AppProfile {
    AppProfile {
        id: "studio".into(),
        name: "Prism Studio".into(),
        version: "0.1.0".into(),
        plugins: None,
        lenses: None,
        default_lens: Some("editor".into()),
        theme: Some(AppThemeConfig {
            display_name: Some("Prism Studio".into()),
            ..Default::default()
        }),
        kbar_commands: None,
        manifest: None,
        allow_glass_flip: Some(true),
        relay_modules: None,
        starter_app: Some(StarterAppTemplate {
            label: "Studio".into(),
            description: Some("Universal host — all lenses, minimal chrome.".into()),
            app_shell: StarterShellSlot {
                data: shell_data([
                    ("brand", json!("Prism Studio")),
                    ("brandIcon", json!("\u{25A6}")),
                    ("topBarHeight", json!(48)),
                    ("leftBarWidth", json!(220)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(true)),
                ]),
                children: Some(vec![
                    shell_child(
                        "heading",
                        "topBar",
                        "Brand",
                        json!({ "text": "Prism Studio", "level": "h3", "align": "left" }),
                    ),
                    shell_child(
                        "site-nav",
                        "leftBar",
                        "Site nav",
                        json!({ "orientation": "vertical", "source": "pages" }),
                    ),
                ]),
            },
            default_page_shell: StarterShellSlot {
                data: shell_data([
                    ("topBarHeight", json!(0)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(false)),
                ]),
                children: None,
            },
            routes: vec![
                route("/", "Home", PageTemplateKind::Landing, true),
                route("/docs", "Docs", PageTemplateKind::Blog, false),
            ],
        }),
    }
}

pub fn flux_profile() -> AppProfile {
    AppProfile {
        id: "flux".into(),
        name: "Flux".into(),
        version: "0.1.0".into(),
        plugins: Some(vec!["work".into(), "finance".into(), "crm".into()]),
        lenses: Some(vec![
            "editor".into(),
            "canvas".into(),
            "record-browser".into(),
            "automation".into(),
            "work".into(),
            "finance".into(),
            "crm".into(),
        ]),
        default_lens: Some("record-browser".into()),
        theme: Some(AppThemeConfig {
            primary: Some("#6C5CE7".into()),
            display_name: Some("Flux".into()),
            brand_icon: Some("flux.svg".into()),
            ..Default::default()
        }),
        kbar_commands: Some(vec![
            "new-task".into(),
            "new-invoice".into(),
            "new-contact".into(),
            "start-timer".into(),
        ]),
        manifest: None,
        allow_glass_flip: Some(true),
        relay_modules: None,
        starter_app: Some(StarterAppTemplate {
            label: "Flux".into(),
            description: Some(
                "Productivity — tasks, contacts, invoices in a record-browser chrome.".into(),
            ),
            app_shell: StarterShellSlot {
                data: shell_data([
                    ("brand", json!("Flux")),
                    ("brandIcon", json!("\u{26A1}")),
                    ("topBarHeight", json!(56)),
                    ("leftBarWidth", json!(280)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(true)),
                ]),
                children: Some(vec![
                    shell_child(
                        "heading",
                        "topBar",
                        "Brand",
                        json!({ "text": "Flux", "level": "h2", "align": "left" }),
                    ),
                    shell_child(
                        "site-nav",
                        "leftBar",
                        "Record nav",
                        json!({ "orientation": "vertical", "source": "pages", "showIcons": true }),
                    ),
                ]),
            },
            default_page_shell: StarterShellSlot {
                data: shell_data([
                    ("topBarHeight", json!(0)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(320)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(false)),
                ]),
                children: None,
            },
            routes: vec![
                route("/", "Inbox", PageTemplateKind::Blank, true),
                route("/tasks", "Tasks", PageTemplateKind::Blank, false),
                route("/contacts", "Contacts", PageTemplateKind::Blank, false),
                route("/invoices", "Invoices", PageTemplateKind::Blank, false),
            ],
        }),
    }
}

pub fn lattice_profile() -> AppProfile {
    AppProfile {
        id: "lattice".into(),
        name: "Lattice".into(),
        version: "0.1.0".into(),
        plugins: Some(vec!["assets".into(), "platform".into()]),
        lenses: Some(vec![
            "editor".into(),
            "graph".into(),
            "spatial-canvas".into(),
            "visual-script".into(),
            "luau-facet".into(),
            "assets-mgmt".into(),
        ]),
        default_lens: Some("graph".into()),
        theme: Some(AppThemeConfig {
            primary: Some("#00B894".into()),
            display_name: Some("Lattice".into()),
            brand_icon: Some("lattice.svg".into()),
            ..Default::default()
        }),
        kbar_commands: Some(vec![
            "new-dialogue".into(),
            "compile-bank".into(),
            "open-entity".into(),
        ]),
        manifest: None,
        allow_glass_flip: Some(true),
        relay_modules: None,
        starter_app: Some(StarterAppTemplate {
            label: "Lattice".into(),
            description: Some("Game middleware — graph-first editor with minimal chrome.".into()),
            app_shell: StarterShellSlot {
                data: shell_data([
                    ("brand", json!("Lattice")),
                    ("brandIcon", json!("\u{2B22}")),
                    ("topBarHeight", json!(40)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(300)),
                    ("bottomBarHeight", json!(120)),
                    ("stickyTopBar", json!(true)),
                ]),
                children: Some(vec![shell_child(
                    "heading",
                    "topBar",
                    "Brand",
                    json!({ "text": "Lattice", "level": "h3", "align": "left" }),
                )]),
            },
            default_page_shell: StarterShellSlot {
                data: shell_data([
                    ("topBarHeight", json!(0)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(false)),
                ]),
                children: None,
            },
            routes: vec![
                route("/", "Graph", PageTemplateKind::Blank, true),
                route("/assets", "Assets", PageTemplateKind::Blank, false),
                route("/scripts", "Scripts", PageTemplateKind::Blank, false),
            ],
        }),
    }
}

pub fn cadence_profile() -> AppProfile {
    AppProfile {
        id: "cadence".into(),
        name: "Cadence".into(),
        version: "0.1.0".into(),
        plugins: Some(vec!["life".into(), "platform".into()]),
        lenses: Some(vec![
            "editor".into(),
            "canvas".into(),
            "luau-facet".into(),
            "record-browser".into(),
            "life".into(),
        ]),
        default_lens: Some("canvas".into()),
        theme: Some(AppThemeConfig {
            primary: Some("#FD79A8".into()),
            display_name: Some("Cadence".into()),
            brand_icon: Some("cadence.svg".into()),
            ..Default::default()
        }),
        kbar_commands: Some(vec![
            "new-lesson".into(),
            "transcribe-session".into(),
            "open-course".into(),
        ]),
        manifest: None,
        allow_glass_flip: Some(true),
        relay_modules: None,
        starter_app: Some(StarterAppTemplate {
            label: "Cadence".into(),
            description: Some("Music education — lesson canvas with hero landing.".into()),
            app_shell: StarterShellSlot {
                data: shell_data([
                    ("brand", json!("Cadence")),
                    ("brandIcon", json!("\u{266B}")),
                    ("topBarHeight", json!(72)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(true)),
                ]),
                children: Some(vec![shell_child(
                    "heading",
                    "topBar",
                    "Brand",
                    json!({ "text": "Cadence", "level": "h1", "align": "center" }),
                )]),
            },
            default_page_shell: StarterShellSlot {
                data: shell_data([
                    ("topBarHeight", json!(0)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(false)),
                ]),
                children: None,
            },
            routes: vec![
                route("/", "Welcome", PageTemplateKind::Landing, true),
                route("/lessons", "Lessons", PageTemplateKind::Blank, false),
                route("/courses", "Courses", PageTemplateKind::Blank, false),
                route("/about", "About", PageTemplateKind::Blog, false),
            ],
        }),
    }
}

pub fn grip_profile() -> AppProfile {
    AppProfile {
        id: "grip".into(),
        name: "Grip".into(),
        version: "0.1.0".into(),
        plugins: Some(vec!["work".into(), "assets".into(), "platform".into()]),
        lenses: Some(vec![
            "editor".into(),
            "graph".into(),
            "spatial-canvas".into(),
            "canvas".into(),
            "automation".into(),
            "work".into(),
            "assets-mgmt".into(),
        ]),
        default_lens: Some("spatial-canvas".into()),
        theme: Some(AppThemeConfig {
            primary: Some("#E17055".into()),
            display_name: Some("Grip".into()),
            brand_icon: Some("grip.svg".into()),
            ..Default::default()
        }),
        kbar_commands: Some(vec![
            "new-cue".into(),
            "open-stage-plot".into(),
            "arm-transport".into(),
        ]),
        manifest: None,
        allow_glass_flip: Some(true),
        relay_modules: None,
        starter_app: Some(StarterAppTemplate {
            label: "Grip".into(),
            description: Some("Live production — cue list driven spatial canvas.".into()),
            app_shell: StarterShellSlot {
                data: shell_data([
                    ("brand", json!("Grip")),
                    ("brandIcon", json!("\u{2726}")),
                    ("topBarHeight", json!(48)),
                    ("leftBarWidth", json!(320)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(160)),
                    ("stickyTopBar", json!(true)),
                ]),
                children: Some(vec![shell_child(
                    "heading",
                    "topBar",
                    "Brand",
                    json!({ "text": "Grip", "level": "h2", "align": "left" }),
                )]),
            },
            default_page_shell: StarterShellSlot {
                data: shell_data([
                    ("topBarHeight", json!(0)),
                    ("leftBarWidth", json!(0)),
                    ("rightBarWidth", json!(0)),
                    ("bottomBarHeight", json!(0)),
                    ("stickyTopBar", json!(false)),
                ]),
                children: None,
            },
            routes: vec![
                route("/", "Stage", PageTemplateKind::Blank, true),
                route("/cues", "Cues", PageTemplateKind::Blank, false),
                route("/assets", "Assets", PageTemplateKind::Blank, false),
            ],
        }),
    }
}

pub fn relay_profile() -> AppProfile {
    AppProfile {
        id: "relay".into(),
        name: "Prism Relay".into(),
        version: "0.1.0".into(),
        plugins: Some(Vec::new()),
        lenses: Some(Vec::new()),
        default_lens: None,
        theme: Some(AppThemeConfig {
            display_name: Some("Prism Relay".into()),
            ..Default::default()
        }),
        kbar_commands: None,
        manifest: None,
        allow_glass_flip: Some(false),
        relay_modules: Some(vec![
            "blind-mailbox".into(),
            "relay-router".into(),
            "relay-timestamp".into(),
            "capability-tokens".into(),
            "sovereign-portals".into(),
            "webrtc-signaling".into(),
        ]),
        starter_app: None,
    }
}

pub fn list_builtin_profiles() -> Vec<AppProfile> {
    vec![
        studio_profile(),
        flux_profile(),
        lattice_profile(),
        cadence_profile(),
        grip_profile(),
        relay_profile(),
    ]
}

pub fn get_builtin_profile(id: BuiltInProfileId) -> AppProfile {
    match id {
        BuiltInProfileId::Studio => studio_profile(),
        BuiltInProfileId::Flux => flux_profile(),
        BuiltInProfileId::Lattice => lattice_profile(),
        BuiltInProfileId::Cadence => cadence_profile(),
        BuiltInProfileId::Grip => grip_profile(),
        BuiltInProfileId::Relay => relay_profile(),
    }
}

#[derive(Debug, thiserror::Error)]
pub enum ProfileParseError {
    #[error("app profile is not valid JSON: {0}")]
    Json(#[from] serde_json::Error),
    #[error("app profile must be a JSON object")]
    NotObject,
    #[error("app profile requires id/name/version strings")]
    MissingFields,
}

pub fn serialize_app_profile(profile: &AppProfile) -> String {
    let mut out = serde_json::to_string_pretty(profile).expect("profile serialises cleanly");
    out.push('\n');
    out
}

pub fn parse_app_profile(source: &str) -> Result<AppProfile, ProfileParseError> {
    let value: JsonValue = serde_json::from_str(source)?;
    let obj = value.as_object().ok_or(ProfileParseError::NotObject)?;
    let has_id = obj.get("id").and_then(|v| v.as_str()).is_some();
    let has_name = obj.get("name").and_then(|v| v.as_str()).is_some();
    let has_version = obj.get("version").and_then(|v| v.as_str()).is_some();
    if !(has_id && has_name && has_version) {
        return Err(ProfileParseError::MissingFields);
    }
    let profile: AppProfile = serde_json::from_value(value)?;
    Ok(profile)
}
